mlbuffer-exploration.md

```diff
@@ -0,0 +1,395 @@
+  may not be as closely tied to a given physical device as an
+  [`IDMLDevice`](https://learn.microsoft.com/en-us/windows/win32/api/directml/nn-directml-idmldevice)?
+  See [#350](https://github.com/webmachinelearning/webnn/issues/350)
+- How will errors be surfaced? See
+  [#477](https://github.com/webmachinelearning/webnn/issues/477). Do we need a
+  concept similar to [WebGPU's error
+  scopes](https://www.w3.org/TR/webgpu/#error-scopes)?
+- Must an `MLBuffer` only be used with an `MLContext` it was created from?
+  (or `MLGraph`s created from that `MLContext`, and so forth)
+- If what we're building is a device-agnostic buffer, it will surely be used for
+  things other than ML (in the long run). In the spirit of
+  [future-proofing](https://w3ctag.github.io/design-principles/#naming-future-proofing),
+  should we name it something other than `MLBuffer`?
+
+## Use Case: Chained Inference
+
+Here's a code sample showing how `MLBuffer`s can be used for chained inference
+and then read back to an `ArrayBuffer`:
+
+```js
+// Create new MLBuffers to be used for chained inference.
+const inputMlBuffer = mlContext.createBuffer({inputSize});
+const intermediateMlBuffer = mlContext.createBuffer({intermediateSize});
+const outputMlBuffer = mlContext.createBuffer({outputSize});
+
+// Copy the contents of an ArrayBuffer into an MLBuffer, to be later used as inputs.
+mlContext.writeBuffer(
+    inputMlBuffer,
+    /*dstOffset=*/0,
+    /*srcData=*/someJsArrayBuffer,
+);
+
+// Perform some ✧*✧* machine learning *✧*✧ described by `graph`.
+mlContext.dispatch(
+    graph,
+    /*inputs=*/{buffer: inputMlBuffer},
+    /*outputs=*/{buffer: intermediateMlBuffer},
+);
+
+// Feed the output of one execution as the input to the next. Chained inference!
+mlContext.dispatch(
+    anotherGraph,
+    /*inputs=*/{buffer: intermediateMlBuffer},
+    /*outputs=*/{buffer: outputMlBuffer},
+);
+
+// Read back the results to script.
+const resultBuffer = await outputMlBuffer.mapAsync();
+```
+
+Let's dive into what happens at each of these steps:
+
+### `MLBuffer` creation
+
+```js
+const inputMlBuffer = mlContext.createBuffer({inputSize});
+```
+#### How it works:
+
+- Enqueue a request on some WebNN timeline to allocate memory on the device
+  associated with `mlContext`
+- The memory allocation will be zeroed (as it is for [WebGPU's `createBuffer()`
+  method](https://www.w3.org/TR/webgpu/#dom-gpudevice-createbuffer))
+
+#### Questions:
+
+- Can an `MLBuffer`'s size always be known at the time of buffer allocation?
+  - In this case and many other cases it seems possible; it's presumably a
+    function of the model and/or video input. But since WebNN always rents a
+    buffer to WebGPU - never the other way around - this introduces a constraint
+    that the size of an `MLBuffer` must always be known at the time of buffer
+    allocation
+- When will `inputMlBuffer` be deallocated if `destroy()` is not called?
+
```

Comment:
I think the main question is whether we can keep these details implementation specific, and use MLBuffer to _identify_ the buffers we pass between stages, eventually with dynamic labels such as input/output/intermediate? (in this case we will have to specify behavior in the spec in more details).

Reply:
> Should we consider create MLBuffer based on its usage?

"usage" can mean a lot of different things (readable/writable? input/output/intermediate? usable by WebGPU? mappable in some way?) but in broad strokes, yes I agree :P

We should strive to ensure the user agent has enough information when an `MLBuffer` is created to allocate its buffer in the most optimal place, according to how it will be used. It's worth enumerating those uses in detail. The table is a very useful start - thanks!

> I think the main question is whether we can keep these details implementation specific

IMHO it's critical that these details are opaque to the website. If creating an "input" `MLBuffer` for the XNNPack backend, the website should not be able to detect that `XNN_EXTRA_BYTES` have been added. For example:
```
const mlBuffer = mlContext.createBuffer({ usage: INPUT, size: 100 });
console.assert(mlBuffer.size, 100);
```
Similarly:
```
const gpuBuffer = mlBuffer.mapAsGpuBuffer(gpuDevice);
console.assert(gpuBuffer.size, mlBuffer.size);

// ...

const arrayBuffer = mlBuffer.mapAsync();
console.assert(arrayBuffer.byteLength, mlBuffer.size);
```

> eventually with dynamic labels such as input/output/intermediate

Could you elaborate more on this? :)